sql: add built-in functions for workload level index recommendations

qiyanghe1998 opened the pull request:
#### sql: add built-in functions for workload level index recommendations

This commit adds 4 generator built-in functions for workload level index
recommendations: workload_index_recs(), workload_index_recs(timestamptz:
TimestampTZ), workload_index_recs(budget: string) and
workload_index_recs(timestamptz: TimestampTZ, budget: string). The return types
of them are all a string column with multiple rows, one row for each index
recommendation.  The user can issue a command like `select
workload_index_recs(...);` and then get the index recommendations for the whole
workload.

Only those workload after the given timestamptz will be considered for the
index recommendations. If there is no timestamptz, all the workload stored in
`system.statement_statistics` will be taken into account. As for the budget, it
represents the space limit for the recommended indexes, e.g. 42GB, 580MiB.  If
it is specified, the workload index recommendation engine will prune the size
of all the indexes to satisfy the space constraint.

A method called workloadindexrec.FindWorkloadRecs() will be implemented to
return the array of indexRecs and error message for these 4 generator built-in
functions to invoke.

Release note: None

Epic: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/105017)
<!-- Reviewable:end -->

qiyanghe1998 committed:
sql: add built-in functions for workload level index recommendations

This commit adds 4 generator built-in functions for workload level index
recommendations: workload_index_recs(), workload_index_recs(timestamptz:
TimestampTZ), workload_index_recs(budget: string) and
workload_index_recs(timestamptz: TimestampTZ, budget: string). The return types
of them are all a string column with multiple rows, one row for each index
recommendation.  The user can issue a command like `select
workload_index_recs(...);` and then get the index recommendations for the whole
workload.

Only those workload after the given timestamptz will be considered for the
index recommendations. If there is no timestamptz, all the workload stored in
`system.statement_statistics` will be taken into account. As for the budget, it
represents the space limit for the recommended indexes, e.g. 42GB, 580MiB.  If
it is specified, the workload index recommendation engine will prune the size
of all the indexes to satisfy the space constraint.

A method called workloadindexrec.FindWorkloadRecs() will be implemented to
return the array of indexRecs and error message for these 4 generator built-in
functions to invoke.

Release note: None

mgartner reviewed (commented):
It might be better to return a string column with multiple rows, one row for each index recommendation, than an array of strings? What do you think? You can look at set-returning built-ins for inspiration: https://github.com/cockroachdb/cockroach/blob/42634711a5869bad8cde5677176cdec7a8f507d0/pkg/sql/sem/builtins/generator_builtins.go

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/105017#-:-NYcUh6d1k02W2NZZT3P:blmct21)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @DrewKimball)



<!-- Sent from Reviewable.io -->

qiyanghe1998 commented:
Yeah, it will be more clear using multiple rows.





<!-- Sent from Reviewable.io -->

mgartner reviewed (commented):
> Reviewed 2 of 5 files at r2.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/105017)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @DrewKimball and @qiyanghe1998)

___
*[`pkg/sql/sem/builtins/generator_builtins_test.go` line 86 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/105017#-NYsz6eH0Jn65UpLbYfF:-NYsz6eH0Jn65UpLbYfG:b-or3v7g) ([raw file](https://github.com/cockroachdb/cockroach/blob/626c990e1217b66371ba282068b86c5d828dbaed/pkg/sql/sem/builtins/generator_builtins_test.go#L86)):*
> ```Go
> 		require.Equalf(t, false, rows.Next(), "failed test case %d", i+1)
> 	}
> }
> ```

It'd be more consistent with the rest of the tests if these were logic tests. Is there a reason not to put them in a logic test like `pkg/sql/logictest/testdata/logic_test/workload_recs`?


<!-- Sent from Reviewable.io -->

qiyanghe1998 commented:
*[`pkg/sql/sem/builtins/generator_builtins_test.go` line 86 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/105017#-NYsz6eH0Jn65UpLbYfF:-NYtZUWl4y3iAWT1hMEb:ba70oie) ([raw file](https://github.com/cockroachdb/cockroach/blob/626c990e1217b66371ba282068b86c5d828dbaed/pkg/sql/sem/builtins/generator_builtins_test.go#L86)):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

It'd be more consistent with the rest of the tests if these were logic tests. Is there a reason not to put them in a logic test like `pkg/sql/logictest/testdata/logic_test/workload_recs`?
</blockquote></details>

You are right. Fixed.


<!-- Sent from Reviewable.io -->

rytaft reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> sorry you've been waiting so long for a review! Just a few nits below.

> Reviewed 1 of 5 files at r2, 11 of 11 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/105017#-:-N_pMfPE72b2_FpfcXeC:b-vtcz97)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @DrewKimball, @mgartner, and @qiyanghe1998)

___
*[`-- commits` line 24 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/105017#-N_pKyiq4xjFdqkwtHhO):*
I think this should have a release note, since these functions will be user-facing
___
*[`pkg/sql/sem/builtins/generator_builtins.go` line 285 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/105017#-N_pLLiW87jsxdWoj0Fw:-N_pLLiW87jsxdWoj0Fx:bk867pn) ([raw file](https://github.com/cockroachdb/cockroach/blob/6bf494c9641bac2d0f1121ed981546ec076a0315/pkg/sql/sem/builtins/generator_builtins.go#L285)):*
> ```Go
> 			tree.ParamTypes{},
> 			types.String,
> 			makeWorkloadIndexRecsGeneratorFactory(false, false),
> ```

nit: for each of these boolean parameters, we usually add a comment to say what it is. Something like this (here and below):
```
makeWorkloadIndexRecsGeneratorFactory(false /* hasTs */, false /* hasBgt */),
```
___
*[`pkg/sql/sem/builtins/generator_builtins.go` line 1121 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/105017#-N_pMRjRDVQSSm3BIIwR:-N_pMRjRDVQSSm3BIIwS:b-lm8hig) ([raw file](https://github.com/cockroachdb/cockroach/blob/6bf494c9641bac2d0f1121ed981546ec076a0315/pkg/sql/sem/builtins/generator_builtins.go#L1121)):*
> ```Go
> // makeWorkloadIndexRecsGeneratorFactory uses the arrayValueGenerator to
> // return all the index recommendations as an array of strings
> func makeWorkloadIndexRecsGeneratorFactory(hasTs bool, hasBgt bool) eval.GeneratorOverload {
> ```

nit: I'd spell these params out: `hasTimestamp` and `hasBudget`. Also update the function comment to mention what these params mean.
___
*[`pkg/sql/sem/builtins/generator_builtins.go` line 1123 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/105017#-N_pLje-CDkyVxaws_vv:-N_pLje-CDkyVxaws_vw:bsgffbo) ([raw file](https://github.com/cockroachdb/cockroach/blob/6bf494c9641bac2d0f1121ed981546ec076a0315/pkg/sql/sem/builtins/generator_builtins.go#L1123)):*
> ```Go
> func makeWorkloadIndexRecsGeneratorFactory(hasTs bool, hasBgt bool) eval.GeneratorOverload {
> 	return func(_ context.Context, _ *eval.Context, _ tree.Datums) (eval.ValueGenerator, error) {
> 		// Invoke the workloadindexrec.FindWorkloadRecs() to get indexRecs, err once it is implemented.
> ```

nit: comment should be <= 80 characters wide. you can just wrap the text here.
___
*[`pkg/sql/logictest/testdata/logic_test/workload_indexrecs` line 1 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/105017#-N_pM3pP5gbi0mxhoXDY:-N_pM3pP5gbi0mxhoXDZ:b53nlrq) ([raw file](https://github.com/cockroachdb/cockroach/blob/6bf494c9641bac2d0f1121ed981546ec076a0315/pkg/sql/logictest/testdata/logic_test/workload_indexrecs#L1)):*
> ```
> # get workload index-recs
> ```

It would be good to add a comment to the top of this file mentioning that the data returned right now is just dummy data.


<!-- Sent from Reviewable.io -->

mgartner commented:
*[`-- commits` line 24 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/105017#-N_pKyiq4xjFdqkwtHhO):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

I think this should have a release note, since these functions will be user-facing
</blockquote></details>

Should we wait on the release note until they are implemented? They're just stubs in this PR.


<!-- Sent from Reviewable.io -->

rytaft reviewed (approved):
> Reviewed 2 of 2 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/105017)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @DrewKimball and @mgartner)

___
*[`-- commits` line 24 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/105017#-N_pKyiq4xjFdqkwtHhO):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

Should we wait on the release note until they are implemented? They're just stubs in this PR.
</blockquote></details>

Fine with me, as long as we don't forget to add one eventually (I think the docs team usually does some massaging anyway, so it's also fine to add something in both places)


<!-- Sent from Reviewable.io -->

qiyanghe1998 commented:
bors r+

yuzefovich commented:
I think this PR is missing `./dev gen docs`.

bors r-

qiyanghe1998 commented:
bors r+
